### PR TITLE
Add ListBuilder::append_value (#3949)

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Part of #3949
Relates to #1841 

# Rationale for this change
 

The current `ListBuilder` API involves appending values to the `ListBuilder::values()` and then finishing the slot with `ListBuilder::append`. Whilst flexible, this API is rather verbose. Now that all the builders implement `Extend` it is possible to provide a more ergonomic API.

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-array/src/builder/generic_list_builder.rs [hunk 1]
```diff
@@ -125,11 +125,97 @@ where
...
+    /// let values = array.values().as_primitive::<Int32Type>();
+    /// assert_eq!(values, &Int32Array::from(vec![Some(1), Some(2), Some(3), None]));
+    /// ```
+    #[inline]
+    pub fn append_value<I, V>(&mut self, i: I)
+    where
+        T: Extend<Option<V>>,
+        I: IntoIterator<Item = Option<V>>,
+    {
+        self.extend(std::iter::once(Some(i)))
```
Comment (tustvold) [inline arrow-array/src/builder/generic_list_builder.rs:194]: At least in some basic experiments LLVM seems to be able to optimise this as one would hope

https://rust.godbolt.org/z/zcxzqzf59

File: arrow-array/src/builder/generic_list_builder.rs [hunk 2]
```diff
@@ -125,11 +125,97 @@ where
...
+    /// Append a value to this [`GenericListBuilder`]
+    ///
+    /// ```
+    /// # use arrow_array::builder::{Int32Builder, ListBuilder};
+    /// # use arrow_array::cast::AsArray;
+    /// # use arrow_array::{Array, Int32Array};
+    /// # use arrow_array::types::Int32Type;
+    /// let mut builder = ListBuilder::new(Int32Builder::new());
+    ///
+    /// builder.append_value([Some(1), Some(2), Some(3)]);
```
Comment (tustvold) [inline arrow-array/src/builder/generic_list_builder.rs:151]: I think this is much easier to read than the current API

File: arrow-array/src/builder/generic_list_builder.rs [hunk 3]
```diff
@@ -125,11 +125,97 @@ where
...
+    #[inline]
+    pub fn append_value<I, V>(&mut self, i: I)
+    where
+        T: Extend<Option<V>>,
+        I: IntoIterator<Item = Option<V>>,
+    {
+        self.extend(std::iter::once(Some(i)))
+    }
+
+    /// Append a null to this [`GenericListBuilder`]
```
Comment (alamb) [inline arrow-array/src/builder/generic_list_builder.rs:197]: I recommend adding a link to `append_value` for an example

```suggestion
    /// Append a null to this [`GenericListBuilder`]
    ///
    /// See [`Self::append_value`] for an example use.
```

**Review by alamb — APPROVED** [review]
I agree -- this is much nicer @tustvold  -- thank you

**Pull request merged**